sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -600,3 +604,146 @@ case class Logarithm(left: Expression, right: Expression)
+        }
+      case _ =>
+        if (scale.dataType.isInstanceOf[IntegralType] && scale.foldable) {
+          // TODO: How to check out of range for foldable LongType Expression
+          // satisfy requirement
+        } else {
+          return TypeCheckFailure("Only foldable Integral Expression " +
+            s"is allowed for ROUND scale arguments, got ${child.dataType}")
+        }
+    }
+    TypeCheckSuccess
+  }
+
+  private lazy val scaleV = scale.eval(EmptyRow)
+  private lazy val _scale = if (scaleV != null) scaleV.asInstanceOf[Int] else 0
+
+  override def eval(input: InternalRow): Any = {
+    val evalE = child.eval(input)
+    if (evalE == null || scaleV == null) return null
+    round(evalE)
+  }
+
+  private lazy val round: (Any) => (Any) = typedRound(child.dataType)
+
+  // Using dataType info to find an appropriate round method
+  private def typedRound(dt: DataType)(x: Any): Any = {
+    dt match {
+      case _: DecimalType =>
+        val decimal = x.asInstanceOf[Decimal]
+        if (decimal.changePrecision(decimal.precision, _scale)) decimal else null
+      case ByteType =>
+        numericRound(x.asInstanceOf[Byte], _scale)
+      case ShortType =>
+        numericRound(x.asInstanceOf[Short], _scale)
+      case IntegerType =>
+        numericRound(x.asInstanceOf[Int], _scale)
+      case LongType =>
+        numericRound(x.asInstanceOf[Long], _scale)
+      case FloatType =>
+        numericRound(x.asInstanceOf[Float], _scale)
+      case DoubleType =>
+        numericRound(x.asInstanceOf[Double], _scale)
+    }
+  }
+
+  private def numericRound[T](input: T, scale: Int)(implicit bdc: BigDecimalConverter[T]): T = {
+    input match {
+      case f: Float if (f.isNaN || f.isInfinite) => return input
+      case d: Double if (d.isNaN || d.isInfinite) => return input
+      case _ =>
+    }
+    bdc.fromBigDecimal(bdc.toBigDecimal(input).setScale(scale, BigDecimal.RoundingMode.HALF_UP))
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val ce = child.gen(ctx)
+
+    def round(primitive: String, integral: Boolean): String = {
+      val (p1, p2) = if (integral) ("new", "") else ("", ".valueOf")
+      s"""
+      ${ev.primitive} = $p1 java.math.BigDecimal$p2(${primitive}).
+        setScale(${_scale}, java.math.BigDecimal.ROUND_HALF_UP)"""
+    }
+
+    def fractionalCheck(primitive: String, function: String): String = {
+      s"""
+      if (Double.isNaN(${primitive}) || Double.isInfinite(${primitive})){
+        ${ev.primitive} = ${primitive};
+      } else {
+        ${round(primitive, false)}.${function};
+      }"""
+    }
+
+    def decimalRound(): String = {
+      s"""
+      if (${ce.primitive}.changePrecision(${ce.primitive}.precision(), ${_scale})) {
+        ${ev.primitive} = ${ce.primitive};
+      } else {
+        ${ev.isNull} = true;
+      }
+      """
+    }
+
+    val roundCode = child.dataType match {
```

Comment:
I'd do something like this to make it simpler

```scala
if (child.dataType.isInstanceOf[IntegralType]) {
  // put the byte/short/.../long handling inline here
} else if (child.dataType.isInstanceOf[DecimalType]) {
  // put the decimal type handling inline here
} else if (child.dataType.isInstanceOf[FractionalType]) {
  // put the float/double handling inline here
}
```

The other thing I'd consider is to special case for scale == 0, since it is very common and much faster to run.